Add parent-aggregation to parent-join module

centic9 opened the pull request:
This is a first try at implementing the parent-aggregation as part of the parent-join module. See #9705 for some previous discussion about this.

This is probably not fully merge-able yet, I would like to get feedback on a few points that I could not clarify myself:
* For implementation I mostly do the same as the child-aggregation does with the main difference of only reporting each parent-document once as part of a bucket
* Does this approach look feasible? It seems to work for our local use-cases, but there might be edge-cases that we do not trigger or detect right now.
* Is there a way to do the Map<Integer, Set<Long>> in ChildrenToParentAggregator.doPostCollection() in a better way? I expect this will create quite some memory-churn. Something like IntMap<LongSet> from Trove or HPPC would be better, but wasn't sure if I can add a dependency on HPPC to the module.
* new unit-tests and all other existing tests do pass, is there any other integration-style testing that we should add?

elasticmachine commented:
Pinging @elastic/es-search-aggs

Kamapcuc commented:
hooray!

jimczi commented on modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregator.java:
```diff
@@ -0,0 +1,216 @@
+            // of the implementation idea, probably needs to be done differently
+            // for production use so we do not cause memory churn here
+            //Set<Tuple<Integer,Long>> seenParents = new HashSet<>();
+            Map<Integer, Set<Long>> seenParents = new HashMap<>();
```
I don't think this is needed because we can have only one parent per global ordinal ? The loop below iterates over the matching parent so you shouldn't see the same global ordinal more than once.

centic9 replied:
Somehow I needed it to get correct results, but it seems I can remove it as soon as I do a check for existing buckets in the LeafCollector.

jimczi commented on modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregator.java:
```diff
@@ -0,0 +1,216 @@
+ *
+ * It ensures that each parent only matches once per bucket.
+ */
+public class ChildrenToParentAggregator extends BucketsAggregator implements SingleBucketAggregator {
```
This class is almost identical to `ParentToChildrenAggregator` so I wonder if they could share a base class that implements all the logic ? It seems that switching `childFilter` and `parentFilter` in `ParentToChildrenAggregator` is the only change here.

jimczi commented on modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregator.java:
```diff
@@ -0,0 +1,216 @@
+                            if (bucketOrds != null) {
+                                bucketOrds = Arrays.copyOf(bucketOrds, bucketOrds.length + 1);
+                                bucketOrds[bucketOrds.length - 1] = bucket;
+                                childrenOrdToOtherBuckets.put(globalOrdinal, bucketOrds);
```
Since we are on the children level we can see the same `bucket` multiple times for the same parent. If it's the same bucket you don't need to switch to an array...

centic9 replied:
You are right, that should make it even faster and lighter on memory. 
If I need to switch to the array, do I need to check if the bucket is already in the array to avoid reporting parents multiple times later?

centic9 commented:
Refactored to have an abstract class for similar code so the two aggregations only contain the different parts.
For Child-To-Parent I now also do a check on the existing bucket before switching to an array and thus could remove the costly lookup in the Map.

jimczi commented on modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregator.java:
```diff
@@ -0,0 +1,102 @@
+                        } else if (bucketPrev != bucket) {
+                            // otherwise store it in the array, but only if the bucket is different than
+                            // the previous one for this childDoc, as we only need one occurrence
+                            // of child -> parent for this aggregation
```
As explained above this part is problematic. It is required to deduplicate the bucket per global ordinal but the cost to do this seems prohibitive. I discussed with @martijnvg offline and he proposed to restrict this new aggregation to the root level (would throw an exception if a parent aggregation is defined). I like this idea because we can assume the same cost for the `parent => children` and `children => parent` cases, we just need an array of global ordinals and `bucket` can be ignored completely.

centic9 replied:
You mean the previous version (https://github.com/elastic/elasticsearch/pull/34210/commits/b6e186cdea45e0987082ba4328e38f93898c8f39#diff-e268a24051f30621c407964faaa1fb24) was correct and this one is not? I verified that all cases that I could think of were handled correctly. Can you sketch out what constellation of aggregations/data would not work here so I can create a test case which verifies it?

My use-case calls for a general implementation as we would like to provide a very flexible search functionality on top of it, but having it enabled for certain uses would be a start. We also experimented with adding a custom plugin/module with the aggregation, so we have a way to go forward with something while having Elasticsearch 6.x in use, so this PR is more a long-term way to integrate this into Elasticsearch itself.

On performance, some rough tests did not show a huge performance problem with this aggregation, but naturally your view of requirements here is much broader, which cases did you think of that will cause issues?

centic9 commented:
Thanks for all this preparatory work! I have now updated the parent-aggregation to make use of the new code. Mostly removing code, test still pass locally.

jimczi commented on modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregator.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Map;
+
+/**
+ * A {@link org.elasticsearch.search.aggregations.bucket.BucketsAggregator} which resolves
```
Can you include `BucketsAggregator` and remove the package for readability ?

jimczi commented on modules/parent-join/src/test/java/org/elasticsearch/join/aggregations/ParentIT.java:
```diff
@@ -0,0 +1,211 @@
+            Parent childrenBucket = categoryBucket.getAggregations().get("to_article");
+            assertThat(childrenBucket.getName(), equalTo("to_article"));
+
+            // TODO: verify some more
```
What is missing here to remove the TODO ?

jimczi commented on modules/parent-join/src/test/java/org/elasticsearch/join/aggregations/ParentIT.java:
```diff
@@ -0,0 +1,211 @@
+        return commenterToComments;
+    }
+
+//    public void testNonExistingParentType() throws Exception {
```
Same here, is this failing ?

centic9 commented:
Hi Jim, I have now fixed the TODOs in the tests, added docu and merged in latest master to resolve a conflict in one test-case, so should be good to go from my side now, let me know if anything is still missing or incorrect :)

jimczi commented:
test this please

jimczi commented:
The failure is not related to this pr. Let's try another one.

jimczi commented:
test this please

centic9 commented:
Perfect, nice that we get this into 6.x as well!
